File: packages/domain/src/weekly-digest.ts
export interface DailyAdherenceRate {
  date: string;
  taken: number;
  total: number;
  rate: number | null;
}

export interface DigestAdherence {
  taken_count: number;
  skipped_count: number;
  total_logged: number;
  completion_rate: number;
  daily_rates: DailyAdherenceRate[];
  best_day: string | null;
  worst_day: string | null;
}

export interface DigestJournal {
  entry_count: number;
  avg_energy: number | null;
  avg_mood: number | null;
  avg_sleep: number | null;
  avg_stress: number | null;
  symptom_frequency: Record<string, number>;
}

export interface DigestExercise {
  session_count: number;
  total_sets: number;
  total_volume: number;
}

export interface DigestMetricDelta {
  current: number | null;
  previous: number | null;
  delta: number | null;
}

export interface WeeklyDigestComparison {
  previous_week_start: string;
  previous_week_end: string;
  adherence_completion_rate: DigestMetricDelta;
  journal_entry_count: DigestMetricDelta;
  journal_avg_energy: DigestMetricDelta;
  exercise_session_count: DigestMetricDelta;
  exercise_total_volume: DigestMetricDelta;
}

export interface MonthlyDigestComparison {
  current_month_start: string;
  current_month_end: string;
  previous_month_start: string;
  previous_month_end: string;
  adherence_completion_rate: DigestMetricDelta;
  journal_entry_count: DigestMetricDelta;
  journal_avg_energy: DigestMetricDelta;
  exercise_session_count: DigestMetricDelta;
  exercise_total_volume: DigestMetricDelta;
}

export interface WeeklyDigest {
  week_start: string;
  week_end: string;
  adherence: DigestAdherence;
  journal: DigestJournal;
  exercise: DigestExercise;
  highlights: string[];
  comparison: WeeklyDigestComparison;
  monthly_comparison: MonthlyDigestComparison;
}

File: apps/api/app/schemas/weekly_digest.py
"""Pydantic schemas for weekly digest."""

from datetime import date

from pydantic import BaseModel


class DailyAdherenceRate(BaseModel):
    date: date
    taken: int
    total: int
    rate: float | None = None


class DigestAdherence(BaseModel):
    taken_count: int
    skipped_count: int
    total_logged: int
    completion_rate: float
    daily_rates: list[DailyAdherenceRate]
    best_day: date | None = None
    worst_day: date | None = None


class DigestJournal(BaseModel):
    entry_count: int
    avg_energy: float | None = None
    avg_mood: float | None = None
    avg_sleep: float | None = None
    avg_stress: float | None = None
    symptom_frequency: dict[str, int] = {}


class DigestExercise(BaseModel):
    session_count: int
    total_sets: int
    total_volume: float


class DigestMetricDelta(BaseModel):
    current: int | float | None = None
    previous: int | float | None = None
    delta: int | float | None = None


class WeeklyDigestComparison(BaseModel):
    previous_week_start: date
    previous_week_end: date
    adherence_completion_rate: DigestMetricDelta
    journal_entry_count: DigestMetricDelta
    journal_avg_energy: DigestMetricDelta
    exercise_session_count: DigestMetricDelta
    exercise_total_volume: DigestMetricDelta


class MonthlyDigestComparison(BaseModel):
    current_month_start: date
    current_month_end: date
    previous_month_start: date
    previous_month_end: date
    adherence_completion_rate: DigestMetricDelta
    journal_entry_count: DigestMetricDelta
    journal_avg_energy: DigestMetricDelta
    exercise_session_count: DigestMetricDelta
    exercise_total_volume: DigestMetricDelta


class WeeklyDigestResponse(BaseModel):
    week_start: date
    week_end: date
    adherence: DigestAdherence
    journal: DigestJournal
    exercise: DigestExercise
    highlights: list[str]
    comparison: WeeklyDigestComparison
    monthly_comparison: MonthlyDigestComparison
